README.rst
*********
Spacewalk
*********

Introduction
============

Spacewalk is an add-on to the `ZeroG`_ job-processing system that auto-generates a discoverable, self-documenting REST API for a set of ZeroG job classes. Pass Spacewalk a modules directory and a base job class derived from ``spacewalk.BaseJob``, and Spacewalk will create a REST API with endpoints for all subclasses of the base job that it finds.

.. _Zerog: https://github.com/tiptapinc/zerog

Install Spacewalk
=================

.. code-block:: console

    $ pip install -e git+https://github.com/tiptapinc/spacewalk.git@0.0.5#egg=spacewalk

Documentation
=============

- ZeroG: https://zerog.readthedocs.io/en/latest/
- Spacewalk: https://spacewalk.readthedocs.io/en/latest

Auto-Generated API
==================

Spacewalk generates a fully discoverable and self-documenting REST API for initiating, monitoring, and querying ZeroG jobs.

Query the base endpoint to get available branches in the endpoint hierarchy:

.. code-block:: console

    $ curl spacewalk:8888/examples/branches

    []

Query a branch to get the available jobs:

.. code-block:: console

    $ curl spacewalk:8888/examples/leaves

    [
        {
            "path": "/examples/waste_time",
            "jobType": "examples_waste_time",
            "name": "Waste Time Job",
            "description": "Randomly logs while wasting time"
        },
        {
            "path": "/examples/fizz_buzz",
            "jobType": "examples_fizz_buzz",
            "name": "FizzBuzz Job",
            "description": "Solve FizzBuzz with settable n, fizz & buzz divisors"
        }
    ]

Query an individual job to get its input parameters in JSON-Schema form:

.. code-block:: console

    $ curl spacewalk:8888/examples/fizz_buzz/post-schema

    {
        "postSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "definitions": {
                "Params": {
                    "properties": {
                        "buzzDivisor": {
                            "title": "buzzDivisor",
                            "type": "number",
                            "format": "integer",
                            "default": 5
                        },
                        "fizzDivisor": {
                            "title": "fizzDivisor",
                            "type": "number",
                            "format": "integer",
                            "default": 3
                        },
                        "n": {
                            "title": "n",
                            "type": "number",
                            "format": "integer",
                            "default": 50
                        }
                    },
                    "type": "object",
                    "additionalProperties": false
                }
            },
            "$ref": "#/definitions/Params"
        }
    }

Use an HTTP POST to initiate a job:

.. code-block:: console

    $ curl -X POST -H 'Content-Type: application/json' -d '{"n":20}' spacewalk:8888/examples/fizz_buzz/job

    {
        "uuid": "d5de4383-ea62-47a0-85fd-419762c457c6"
    }

Monitor the job's progress:

.. code-block:: console

    $ curl spacewalk:8888/examples/progress/d5de4383-ea62-47a0-85fd-419762c457c6

    {
        "completeness": 1.0,
        "result": 200
    }

Get full details of the job's run:

.. code-block:: console

    $ curl spacewalk:8888/examples/info/d5de4383-ea62-47a0-85fd-419762c457c6

    {
        "completeness": 1.0,
        "result": 200,
        "events": [
            {
                "msg": "starting examples_fizz_buzz job d5de4383-ea62-47a0-85fd-419762c457c6",
                "timeStamp": "2021-09-28T14:51:57.209432"
            }
        ],
        "errors": [],
        "warnings": []
    }

Get the job's results:

.. code-block:: console

    $ curl spacewalk:8888/examples/data/d5de4383-ea62-47a0-85fd-419762c457c6

    {
        "output": [
            "1",
            "2",
            "Fizz",
            "4",
            "Buzz",
            "Fizz",
            "7",
            "8",
            "Fizz",
            "Buzz",
            "11",
            "Fizz",
            "13",
            "14",
            "FizzBuzz",
            "16",
            "17",
            "Fizz",
            "19",
            "Buzz"
        ]
    }

Get a full dump of the job (showLogs parameter inludes or excludes logs, false by default):

.. code-block:: console

    $ curl spacewalk:8888/examples/dump/9726b1fa-01e0-4c10-931e-55128415587d?showLogs=true

    {
        "documentType": "zerog_job",
        "datasetName": "bids",
        "endDate": null,
        "createdAt": "2021-11-20T02:55:10.967790",
        "logId": "gaql_load-dataset_9726b1fa-01e0-4c10-931e-55128415587d",
        "startDate": null,
        "queueKwargs": {
            "ttr": 2592000
        },
        "updatedAt": "2021-11-20T02:55:25.709894",
        "name": "Boots Report",
        "schemaVersion": 1.0,
        "running": false,
        "uuid": "9726b1fa-01e0-4c10-931e-55128415587d",
        "queueJobId": 4,
        "customerId": "2464420064",
        "errorCount": 0,
        "jobType": "gaql_load-dataset",
        "tickcount": 1.0,
        "completeness": 1.0,
        "resultCode": 200,
        "cas": 1637376925711663104,
        "tickval": 1.0,
        "action": "load"
}


CHANGELOG.md
## v0.0.9
* remove zerog from requirements.txt

## v0.0.8
* update zerog version in setup.py

## v0.0.7
* add /dump handler for dumping jobs

## v0.0.6
* use zerog version that requires Couchbase SDK v3+
* upgrade documentation

## v0.0.5
* add /info endpoint

## v0.0.4
* add support for ZeroG /data handler

## v0.0.3
* don't create missing Params 

## v0.0.2

* remove unneeded log message

## v0.0.1:

* Initial commit


example.py
from marshmallow import fields, Schema
import random
import spacewalk
import sys
import time
import tornado.ioloop
import zerog

import logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - "
           "%(message)s - [%(process)s:%(name)s:%(funcName)s]"
)
log = logging.getLogger(__name__)

INTERVAL = 5
MAX_TIME = 120


class BaseExampleJobSchema(spacewalk.BaseJobSchema):
    pass


class BaseExampleJob(spacewalk.BaseJob):
    NAME = "Spacewalk Example"
    BRANCH_NAME = "examples"
    DESCRIPTION = "example Spacewalk jobs"

    BASE_SCHEMA = BaseExampleJobSchema

    def run(self):
        self.job_log_info(f"starting {self.jobType} job {self.uuid}")


class WasteTimeJob(BaseExampleJob):
    NAME = "Waste Time Job"
    LEAF_NAME = "waste_time"
    DESCRIPTION = "Randomly logs while wasting time"

    class Params(Schema):
        delay = fields.Integer()

    def run(self):
        super().run()

        end = time.time() + self.delay
        logInterval = self.delay / 10

        while True:
            if time.time() > end:
                break

            logDelay = (random.random() + 0.5) * logInterval
            time.sleep(logDelay)
            self.add_to_completeness(logDelay / self.delay)
            self.job_log_info(f"{end - time.time():.2f} seconds remaining")

        return 200, None


class FizzBuzzJobSchema(BaseExampleJobSchema):
    output = fields.List(fields.String)


class FizzBuzzJob(BaseExampleJob):
    NAME = "FizzBuzz Job"
    LEAF_NAME = "fizz_buzz"
    DESCRIPTION = "Solve FizzBuzz with settable n, fizz & buzz divisors"

    BASE_SCHEMA = FizzBuzzJobSchema

    class Params(Schema):
        n = fields.Integer(default=50, missing=50)
        fizzDivisor = fields.Integer(default=3, missing=3)
        buzzDivisor = fields.Integer(default=5, missing=5)

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.output = kwargs.get("output", [])

    def run(self):
        super().run()

        divisors = [self.fizzDivisor, self.buzzDivisor]
        output = []

        for i in range(1, self.n + 1):
            remainders = list(map(lambda x: i % x, divisors))
            if remainders == [0, 0]:
                msg = "FizzBuzz"
            elif remainders[0] == 0:
                msg = "Fizz"
            elif remainders[1] == 0:
                msg = "Buzz"
            else:
                msg = str(i)

            output.append(msg)

        self.update_attrs(output=output)
        return 200, None

    def get_data(self):
        # start with the base data response and add our own
        data = super().get_data()
        data['output'] = self.output
        return data


def make_datastore():
    return zerog.CouchbaseDatastore(
        "couchbase", "Administrator", "password", "test"
    )


def make_queue(queueName):
    return zerog.BeanstalkdQueue("beanstalkd", 11300, queueName)


def start_server():
    tree = spacewalk.auto_tree(BaseExampleJob, "")
    struct = spacewalk.Structure(tree)
    handlers = spacewalk.make_handlers(struct)

    server = spacewalk.Server(
        struct,
        "myService",
        make_datastore,
        make_queue,
        [WasteTimeJob, FizzBuzzJob],
        handlers
    )
    server.listen(8888)
    tornado.ioloop.IOLoop.current().start()


def datastore_health():
    try:
        make_datastore()
        log.info("datastore is healthy")
        return True
    except BaseException:
        log.info("waiting for datastore")
        return False


def main():
    startTime = time.time()
    while time.time() < startTime + MAX_TIME:
        time.sleep(INTERVAL)

        if datastore_health():
            start_server()

    sys.exit(1)


if __name__ == '__main__':
    main()
